gateway-rest/src/main/java/io/camunda/zeebe/gateway/rest/UserTaskController.java

```diff
@@ -0,0 +1,159 @@
+      final UserTaskCompletionRequest completionRequest,
+      final Long userTaskKey,
+      final ServerWebExchange context) {
+    final CompletableFuture<ResponseEntity<Object>> finalResponse = new CompletableFuture<>();
+
+    final BrokerUserTaskCompletionRequest brokerRequest =
+        mapToBrokerRequest(completionRequest, userTaskKey, context);
+
+    final CompletableFuture<BrokerResponse<UserTaskRecord>> brokerResponse =
+        brokerClient.sendRequest(brokerRequest);
+
+    brokerResponse.whenCompleteAsync(
+        (response, error) -> {
+          var restResponse = ResponseEntity.noContent().build();
+          if (error != null) {
+            restResponse = getErrorResponse(error);
+          } else if (response.getRejection() != null) {
+            restResponse = getRejectionResponse(response.getRejection());
+          }
+          finalResponse.complete(restResponse);
+        });
+
+    return finalResponse;
+  }
+
+  private static BrokerUserTaskCompletionRequest mapToBrokerRequest(
+      final UserTaskCompletionRequest completionRequest,
+      final Long userTaskKey,
+      final ServerWebExchange context) {
+
+    final var brokerRequest =
+        new BrokerUserTaskCompletionRequest(
+            userTaskKey,
+            getDocumentOrEmpty(completionRequest, UserTaskCompletionRequest::getVariables),
+            getStringOrEmpty(completionRequest, UserTaskCompletionRequest::getAction));
+
+    final List<String> authorizedTenants =
+        context.getAttributeOrDefault(
+            TENANT_CTX_KEY, List.of(TenantOwned.DEFAULT_TENANT_IDENTIFIER));
+
+    final String authorizationToken =
+        Authorization.jwtEncoder()
+            .withIssuer(JwtAuthorizationBuilder.DEFAULT_ISSUER)
+            .withAudience(JwtAuthorizationBuilder.DEFAULT_AUDIENCE)
+            .withSubject(JwtAuthorizationBuilder.DEFAULT_SUBJECT)
+            .withClaim(Authorization.AUTHORIZED_TENANTS, authorizedTenants)
+            .encode();
+    brokerRequest.setAuthorization(authorizationToken);
+
+    return brokerRequest;
+  }
+
+  private static DirectBuffer getDocumentOrEmpty(
+      final UserTaskCompletionRequest request,
+      final Function<UserTaskCompletionRequest, Map<String, Object>> variablesExtractor) {
+    final Map<String, Object> value = request == null ? null : variablesExtractor.apply(request);
+    return value == null || value.isEmpty()
+        ? DocumentValue.EMPTY_DOCUMENT
+        : new UnsafeBuffer(MsgPackConverter.convertToMsgPack(value));
+  }
+
+  private static String getStringOrEmpty(
+      final UserTaskCompletionRequest request,
+      final Function<UserTaskCompletionRequest, String> valueExtractor) {
+    final String value = request == null ? null : valueExtractor.apply(request);
+    return value == null ? "" : value;
+  }
+
+  private ResponseEntity<Object> getErrorResponse(final Throwable error) {
+    return ResponseEntity.badRequest()
```

Comment:
:x: The following test fails right now:

```java
  @Test
  public void shouldFailOnInvalidResponseType() {
    // given
    final var request = new UserTaskCompletionRequest();
    brokerResponseFutureSupplier =
        () -> CompletableFuture.completedFuture(new BrokerResponse<>(new JobRecord()));

    // when / then
    webClient
        .post()
        .uri("api/v1/user-tasks/1/completion")
        .accept(MediaType.APPLICATION_JSON)
        .contentType(MediaType.APPLICATION_JSON)
        .body(Mono.just(request), UserTaskCompletionRequest.class)
        .exchange()
        .expectStatus()
        .is5xxServerError()
        .expectBody()
        .isEmpty();
  }
```

Basically, you get a 204 even if the broker responded with a nonsensical response.

> [!Note]
> This is a fake test, I don't expect you to make it pass :smile: It's just how I came across this issue.

Essentially, because we're mocking the response, `BrokerUserTaskCompletionRequest#toResponseDto` is never called. In practice, it would be, and decoding the response will throw an error which I think will result in a `BrokerErrorResponse`, meaning you'll get an exception in the handler in the controller. At the moment, for all errors, we return BAD_REQUEST.

I'm not sure `BAD_REQUEST` is the right status here, not for all of them. We can take a look at `GrpcErrorMapper#mapBrokerErrorToStatus`:

```java
  private Status mapBrokerErrorToStatus(
      final Throwable rootError, final BrokerError error, final Logger logger) {
    final Builder builder = Status.newBuilder();
    String message = error.getMessage();

    switch (error.getCode()) {
      case PROCESS_NOT_FOUND -> builder.setCode(Code.NOT_FOUND_VALUE);
      case RESOURCE_EXHAUSTED -> {
        builder.setCode(Code.RESOURCE_EXHAUSTED_VALUE);
        logger.trace("Target broker is currently overloaded: {}", error, rootError);
      }
      case PARTITION_LEADER_MISMATCH -> {
        // return UNAVAILABLE to indicate to the user that retrying might solve the issue, as this
        // is usually a transient issue
        logger.trace("Target broker was not the leader of the partition: {}", error, rootError);
        builder.setCode(Code.UNAVAILABLE_VALUE);
      }
      case MALFORMED_REQUEST -> builder.setCode(Code.INVALID_ARGUMENT_VALUE);
      default -> {
        // all the following are for cases where retrying (with the same gateway) is not expected
        // to solve anything
        logger.error(
            "Expected to handle gRPC request, but received an internal error from broker: {}",
            error,
            rootError);
        builder.setCode(Code.INTERNAL_VALUE);
        message =
            String.format(
                "Unexpected error occurred between gateway and broker (code: %s) (message: %s)",
                error.getCode(), error.getMessage());
      }
    }

    return builder.setMessage(message).build();
  }
```

I think only `PROCESS_NOT_FOUND` is an actual client error, in a way. The others aren't client errors. Here we should handled them differently, for example:

- RESOURCE_EXHAUSTED should return 429. It's not the most accurate, but closest. We may want to return a body here as well, though I haven't fully thought about it.
- PARTITION_LEADER_MISMATCH is typically transient during an election, or if too many nodes are down. This is typically retry-able, so we should return an error code matching that, like 503.
- MALFORMED_REQUEST means the broker failed to parse the incoming gateway request. There is _nothing_ a client can do here, and this is pretty bad (hence why we also log it as an error on the gRPC side). We should log it as an error (so SREs are notified), and return 500.

Whether we want to do it with the other codes...well, I suppose. Why not. The others are codes that should _never_ be sent, like `NULL_VAL` and `SBE_UNKNOWN`. So we should also return 500 and log it.

Reply:
Awesome, great detail, that helps 👍 I'll have a look 🙂